### PR TITLE
New jupyter notebook examples

### PR DESCRIPTION
**Pull request by SRFU-NN**

### CODE REVIEW

File: examples/new_examples.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+## Multiobjective
+
+1. New train line
+2. Plots
+3. Tradeoffs
+
+## Constraints
+
+1. Cut off "corner"
```
Comment (dk-teknologisk-mon) [inline examples/new_examples.md]: This seems like the right place to add a "2. Constrain factors to a sum".

**Comment by sqbl** [comment]
LGTM. Really good

Small nitty-gritty suggestions from me:

- "Replay" the entire notebook to make the numbers of the cells become 1, 2, 3...
- Typo: Teel -> Tell (in the chapter "starting to dig")
- Typo: put -> our/the(?) (in the chapter "The ProcessOptimizer")
- Typo: 2xTeh -> 2xthe (in the chapter "Stopping here, or moving on")
- Suggested additional feature: I think Ms Dixon (nice touch by the way...) forgot about her dig at 7.5,7.5. It a shame to let that datapoint go to waste. She could tell it to the algorithm as she realized that she had forgotten it after making the initial 10 digs(?)

General question: Doesn't the initial 10 LHS points seem strange? I would have expected another distribution.

**Comment by SRFU-NN** [comment]
Great, thanks!

We should rerun the notebooks later, just before the main merge. I expect there to be some reorganisation before then anyway, with splitting or merging notebooks to make the story flow better.

The LHS points fall suspiciously close to the diagonal, if that is what you see. But they seem to have the correct distribution, one in each 1/10 of each axis, so it seems like just “bad luck”. They have the same distribution regardless of what the random seed is set to.

Best regards
Søren Furbo
Automation Scientist
7660 Automation and Process Optimisation
+45 3079 0427 (direct) ǀ  ***@***.******@***.***>

**Comment by dk-teknologisk-mon** [comment]
About the LHS sampling: The result is actually not surprising if you dive into it a bit further. LHS is specifically designed to provide a perfectly even distribution of points _after projection to one dimension_, not in the original space. This was actually what led me to the quasirandom sequence (QRS) sampling that I implemented as part of the SumEqual constrains. Like the Sobol sequence, it is designed to be very good at space filling in the dimensionality it starts in. 

I created a couple of examples comparing LHS and QRS in 2D, with a color scale indicating distance to nearest data point. The 10 data points is what you see in your notebook example:

Figure_1
Figure_3

The gut reaction to seeing this may be to think "we should drop LHS and use QRS!", but unfortunately that lunch isn't free... The reason is that while LHS is bad at space filling in order to be good at projection, the reverse is also true, namely that QRS is really good at space filling in any dimension but does really strange things when projected to lower dimensions.

To illustrate this, I have generated LHS and QRS points in a 3D and 5D parameter space and projected the points down to 2D:

Figure_3D_to_2D
Figure_5D_to_2D

The severity of the projection problem changes unpredictably depending on which axes you project along for QRS, but LHS is consistent (because it was designed to be so):

Figure_5D_to_2D_v2

So the trade-off we have to make is whether we believe that it is more common that users of this package chuck in many parameters without any effect or not. In the former case, LHS is probably the better initial sampling strategy while QRS probably performs better in the latter case. That said, it probably also matters how "local" the good solution is as I suspect QRS will find "golf hole" solutions much faster than LHS. I suggest we include investigation of these things in the list of things to validate with benchmarking, as using LHS or QRS for initial sampling is a clear hyperparameter.

**Comment by SRFU-NN** [comment]
> About the LHS sampling: The result is actually not surprising if you dive into it a bit further. LHS is specifically designed to provide a perfectly even distribution of points _after projection to one dimension_, not in the original space. This was actually what led me to the quasirandom sequence (QRS) sampling that I implemented as part of the SumEqual constrains. Like the Sobol sequence, it is designed to be very good at space filling in the dimensionality it starts in.
> 
> I created a couple of examples comparing LHS and QRS in 2D, with a color scale indicating distance to nearest data point. The 10 data points is what you see in your notebook example:
> 
> Figure_1 Figure_3
> 
> The gut reaction to seeing this may be to think "we should drop LHS and use QRS!", but unfortunately that lunch isn't free... The reason is that while LHS is bad at space filling in order to be good at projection, the reverse is also true, namely that QRS is really good at space filling in any dimension but does really strange things when projected to lower dimensions.
> 
> To illustrate this, I have generated LHS and QRS points in a 3D and 5D parameter space and projected the points down to 2D:
> 
> Figure_3D_to_2D Figure_5D_to_2D
> 
> The severity of the projection problem changes unpredictably depending on which axes you project along for QRS, but LHS is consistent (because it was designed to be so):
> 
> Figure_5D_to_2D_v2
> 
> So the trade-off we have to make is whether we believe that it is more common that users of this package chuck in many parameters without any effect or not. In the former case, LHS is probably the better initial sampling strategy while QRS probably performs better in the latter case. That said, it probably also matters how "local" the good solution is as I suspect QRS will find "golf hole" solutions much faster than LHS. I suggest we include investigation of these things in the list of things to validate with benchmarking, as using LHS or QRS for initial sampling is a clear hyperparameter.

Really interesting, and great plots to illustrate the issue! Is this generally a problem for QRS's, or could the R<sub>1</sub> sequence you send a link about do better? The projection behaviour looks eerily close to the issues he points out about some of the other QRS's.

**Comment by dk-teknologisk-mon** [comment]
> Really interesting, and great plots to illustrate the issue! Is this generally a problem for QRS's, or could the R1 sequence you send a link about do better? The projection behaviour looks eerily close to the issues he points out about some of the other QRS's.

When I say QRS, I specifically mean the R<sub>1</sub> sequence from that link :) I vaguely recall a discussion on hackernews about the blog where the original author stated that this projection problem is fairly general across spacefilling sequences/algorithms. So no matter if you choose R<sub>1</sub>, Sobol or something else, they will tend to perform poorly/weird in projection to lower dimensions.

**Comment by SRFU-NN** [comment]
@MortenSkovsted: Would you read through `.\examples\new_examples\start_here.ipynb`, `.\examples\new_examples\explore_vs_exploit.ipynb`, and `.\examples\new_examples\plots.ipynb`, and see if they make sense for a person who have just discovered the package?

@dk-teknologisk-mon , @sqbl : In `.\examples\new_examples\batch_vs_single.ipynb`, I don't know what to write after the last code block with the strategies, since I don't have enough experience with them. Do you have any ideas? And please look at `.\examples\new_examples\narrow_wells.ipynb`, and tell me if it should be included, and if yes, what we can write as a conclusion.

**Comment by MortenSkovsted** [comment]
Waw I really like the whole storyline! (examples/new_examples/gold_mining.md) and (examples/new_examples/start_here.ipynb)

First of all sorry for my Github skills, I will try to make a few corrections/suggestions in this comment.

"examples/new_examples/explore_vs_exploit.ipynb" is also easy to understand, and nice names.
Last line need at spell check ...", and to set the axplore-exploit balance accordingly." 


"examples/new_examples/plots.ipynb"
"The coloured plot in the lower left (2D plot) shows the internal model for how much gold is will be found for each position."
"The coloured plot in the lower left (2D plot) shows the internal model for how much gold _the model expects to find?_ ."

I don't get line 6 (Plotting model after 3, 9 and 50 digs), there seem to be way more than 3 and 9 digs.
image


Really great work in general!! I have sent it to a coworker who has even less experience with Bayesian Optimization than me to see how he likes it.

**Comment by sqbl** [comment]
> @MortenSkovsted: Would you read through `.\examples\new_examples\start_here.ipynb`, `.\examples\new_examples\explore_vs_exploit.ipynb`, and `.\examples\new_examples\plots.ipynb`, and see if they make sense for a person who have just discovered the package?
> 
> @dk-teknologisk-mon , @sqbl : In `.\examples\new_examples\batch_vs_single.ipynb`, I don't know what to write after the last code block with the strategies, since I don't have enough experience with them. Do you have any ideas? And please look at `.\examples\new_examples\narrow_wells.ipynb`, and tell me if it should be included, and if yes, what we can write as a conclusion.

Hi,
Regarding the discussion on batch vs single experimentation, we could also add the last available strategy (Kriging believer). This strategy was found to be the optimal strategy [this](https://www.nature.com/articles/s41586-021-03213-y) paper. But that doesn't answer you question on what to write now... I would simply write that the chosen strategy migth/should be chosen based on the type of task at hand for the experimentor (which, I know, is not really an argument at all...). Then you should end the paragraph with a repetition of your points from a few lines above, that batch mode is (emperically) worse than real Bayesian optimization. If anything, you might direct the readers attention to where the digs have been performed in the landscape and, hence, direct the discussion to a explore/exploit discussion (?)

**Comment by sqbl** [comment]
> @MortenSkovsted: Would you read through `.\examples\new_examples\start_here.ipynb`, `.\examples\new_examples\explore_vs_exploit.ipynb`, and `.\examples\new_examples\plots.ipynb`, and see if they make sense for a person who have just discovered the package?
> 
> @dk-teknologisk-mon , @sqbl : In `.\examples\new_examples\batch_vs_single.ipynb`, I don't know what to write after the last code block with the strategies, since I don't have enough experience with them. Do you have any ideas? And please look at `.\examples\new_examples\narrow_wells.ipynb`, and tell me if it should be included, and if yes, what we can write as a conclusion.

I like the narrow wells type idea. I find the though very relevant...
- Do we want to take the story towards a discussion on the search space (are you even searching in the right design-space)?
- Or do we want to start the discussion that we carry no garantee for finding global optimum? Perhaps with a story that the gold-digger party needs to ask themselfes if they can make a business out of digging gold in any of the areas identified after the initial 20 digs(?)

**Review by sqbl — APPROVED** [review]
LGTM

**Pull request merged**